src/mongocxx/collection.cpp

```diff
@@ -327,30 +327,24 @@ stdx::optional<result::delete_result> collection::delete_one(
     return stdx::optional<result::delete_result>(result::delete_result(std::move(result.value())));
 }
 
-stdx::optional<bsoncxx::document::value> collection::find_one_and_replace(
-    bsoncxx::document::view filter, bsoncxx::document::view replacement,
-    const options::find_one_and_replace& options) {
+stdx::optional<bsoncxx::document::value> _find_and_modify(::mongoc_collection_t* collection,
+                                                          bsoncxx::document::view filter,
+                                                          ::mongoc_find_and_modify_opts_t* opts) {
     scoped_bson_t bson_filter{filter};
-    scoped_bson_t bson_replacement{replacement};
-    scoped_bson_t bson_sort{options.sort()};
-    scoped_bson_t bson_projection{options.projection()};
-
     scoped_bson_t reply;
     reply.flag_init();
 
-    bson_error_t error;
+    ::bson_error_t error;
 
-    options::return_document rd =
-        options.return_document().value_or(options::return_document::k_before);
+    bool r = libmongoc::collection_find_and_modify_with_opts(collection, bson_filter.bson(), opts,
+                                                             reply.bson(), &error);
 
-    // TODO handle bypass document validation when CDRIVER-945 is done.
-    bool r = mongoc_collection_find_and_modify(
-        _impl->collection_t, bson_filter.bson(), bson_sort.bson(), bson_replacement.bson(),
-        bson_projection.bson(), false, options.upsert().value_or(false),
-        rd == options::return_document::k_after, reply.bson(), &error);
+    libmongoc::find_and_modify_opts_destroy(opts);
```

Comment:
I don't really like that this function takes a raw pointer to an opts allocated elsewhere and then destroys it. I think I'd prefer that callers manage the lifetime of their own objects. I think you should just have this function take a const ref to the opts_t. Make the callers manage the lifetime.